tools/autograd/gen_variable_type.py

```diff
@@ -80,6 +79,67 @@
     '_coalesced_',
 }
 
+# NOTE [ Invariant: TensorImpl and Storage Pointer Equality ]
+#
+# When a function modifies its input tensors (via inplace or out-variants),
+# it should never change the the input tensors' underlying c10::TensorImpl pointers
+# or c10::Storage pointers.
+#
+# The following code templates implement the checks for this invariant:
+SAVE_TENSOR_STORAGE = CodeTemplate("""\
+Storage ${tensor_name}_storage_saved =
+  ${tensor_name}.has_storage() ? ${tensor_name}.storage() : Storage();
+""")
+
+ENFORCE_SAME_TENSOR_STORAGE = CodeTemplate("""\
+if (${tensor_name}_storage_saved) AT_ASSERT(${tensor_name}_storage_saved.is_alias_of(${tensor_name}.storage()));
+""")
+
+SAVE_TENSORLIST_STORAGE = CodeTemplate("""\
+std::vector<Storage> ${tensorlist_name}_storage_saved(${tensorlist_name}.size());
+for (Tensor tensor : ${tensorlist_name})
+  ${tensorlist_name}_storage_saved.push_back(tensor.has_storage() ? tensor.storage() : Storage());
+""")
+
+ENFORCE_SAME_TENSORLIST_STORAGE = CodeTemplate("""\
+for (size_t i=0; i<${tensorlist_name}.size(); i++) {
```

Comment:
it might be cleaner to just use `for(elem : list)`.

Reply:
We actually can't use `for(elem : list)` here because we are iterating two lists (`${tensorlist_name}_storage_saved` and `${tensorlist_name}`) at the same time.